vendor/github.com/openshift/library-go/pkg/monitor/health/prober.go

```diff
@@ -0,0 +1,377 @@
+		sm.unhealthyTargets = unhealthyTargetsSet.List()
+	}
+
+	sm.targetsToMonitor = freshTargets
+}
+
+func (sm *Prober) healthCheckRegisteredTargets() {
+	sm.refreshTargetsLocked()
+	var wg sync.WaitGroup
+	resTargetErrTupleCh := make(chan targetErrTuple, len(sm.targetsToMonitor))
+
+	for i := 0; i < len(sm.targetsToMonitor); i++ {
+		wg.Add(1)
+		go func(target string) {
+			defer wg.Done()
+			err := sm.healthCheckSingleTarget(target)
+			resTargetErrTupleCh <- targetErrTuple{target, err}
+		}(sm.targetsToMonitor[i])
+	}
+	wg.Wait()
+	close(resTargetErrTupleCh)
+
+	currentHealthCheckProbes := make([]targetErrTuple, 0, len(sm.targetsToMonitor))
+	for svrErrTuple := range resTargetErrTupleCh {
+		currentHealthCheckProbes = append(currentHealthCheckProbes, svrErrTuple)
+	}
+
+	sm.updateHealthChecksFor(currentHealthCheckProbes)
+}
+
+// updateHealthChecksFor examines the health of targets based on the provided probes and the current configuration.
+// It also notifies interested parties about changes in the health condition.
+// Interested parties can be registered by calling AddListener method.
+func (sm *Prober) updateHealthChecksFor(currentHealthCheckProbes []targetErrTuple) {
+	newUnhealthyTargets := []string{}
+	newHealthyTargets := []string{}
+
+	for _, svrErrTuple := range currentHealthCheckProbes {
+		if svrErrTuple.err != nil {
+			delete(sm.consecutiveSuccessfulProbes, svrErrTuple.target)
+
+			unhealthyProbesSlice := sm.consecutiveFailedProbes[svrErrTuple.target]
+			if len(unhealthyProbesSlice) < sm.unhealthyProbesThreshold {
+				unhealthyProbesSlice = append(unhealthyProbesSlice, svrErrTuple.err)
+				sm.consecutiveFailedProbes[svrErrTuple.target] = unhealthyProbesSlice
+				if len(unhealthyProbesSlice) == sm.unhealthyProbesThreshold {
+					newUnhealthyTargets = append(newUnhealthyTargets, svrErrTuple.target)
+				}
+			}
+			continue
+		}
+
+		delete(sm.consecutiveFailedProbes, svrErrTuple.target)
+
+		healthyProbesCounter := sm.consecutiveSuccessfulProbes[svrErrTuple.target]
+		if healthyProbesCounter < sm.healthyProbesThreshold {
+			healthyProbesCounter++
+			sm.consecutiveSuccessfulProbes[svrErrTuple.target] = healthyProbesCounter
+			if healthyProbesCounter == sm.healthyProbesThreshold {
+				newHealthyTargets = append(newHealthyTargets, svrErrTuple.target)
+			}
+		}
+	}
+
+	newUnhealthyTargetsSet := sets.NewString(newUnhealthyTargets...)
+	newHealthyTargetsSet := sets.NewString(newHealthyTargets...)
+	notifyListeners := false
+
+	// detect unhealthy targets
+	previouslyUnhealthyTargetsSet := sets.NewString(sm.unhealthyTargets...)
+	currentlyUnhealthyTargetsSet := previouslyUnhealthyTargetsSet.Union(newUnhealthyTargetsSet)
+	currentlyUnhealthyTargetsSet.Delete(newHealthyTargetsSet.List()...)
+	if !currentlyUnhealthyTargetsSet.Equal(previouslyUnhealthyTargetsSet) {
+		sm.unhealthyTargets = currentlyUnhealthyTargetsSet.List()
+		klog.V(2).Infof("observed the following unhealthy targets %v", sm.unhealthyTargets)
```

Comment:
would be interesting to log if is an increment or a decrement, i.e , we have more or less unhealthy targets?
this will avoid having to scroll down on the logs 
same for healthy ones

Reply:
we do have metrics for knowing which targets are currently healthy/unhealthy